src/main/java/org/apache/maven/plugins/dependency/tree/GraphmlDependencyNodeVisitor.java

```diff
@@ -63,37 +66,55 @@ public GraphmlDependencyNodeVisitor(Writer writer) {
      * {@inheritDoc}
      */
     @Override
-    public boolean endVisit(DependencyNode node) {
-        if (node.getParent() == null || node.getParent() == node) {
-            writer.write(GRAPHML_FOOTER);
-        } else {
-            DependencyNode p = node.getParent();
-            writer.print("<edge source=\"" + generateId(p) + "\" target=\"" + generateId(node) + "\">");
-            if (node.getArtifact().getScope() != null) {
-                // add Edge label
-                writer.print("<data key=\"d1\"><y:PolyLineEdge><y:EdgeLabel>"
-                        + node.getArtifact().getScope() + "</y:EdgeLabel></y:PolyLineEdge></data>");
+    public boolean visit(DependencyNode node) {
+        try {
+            StringWriter stringWriter = new StringWriter();
+            if (node.getParent() == null || node.getParent() == node) {
+                stringWriter.write(GRAPHML_HEADER);
             }
-            writer.println("</edge>");
+            // write node
+            stringWriter.write("<node id=\"" + generateId(node) + "\">");
+            // add node label
+            stringWriter.write("<data key=\"d0\"><y:ShapeNode><y:NodeLabel>" + node.toNodeString()
+                    + "</y:NodeLabel></y:ShapeNode></data>");
+            stringWriter.write("</node>\n");
+
+            // Write the accumulated output to the provided writer
+            writer.write(stringWriter.toString());
+            writer.flush();
+            return true;
+        } catch (IOException e) {
+            throw new UncheckedIOException("Failed to write GraphML node", e);
```

Comment:
I'm starting to see how this should work. Either drop all the stringwiters and write straight onto the writer, or use stringwriters to build up a  report that is written after the visitor is finished.